### PR TITLE
New data loading: Save a mid epoch checkpoint every x minutes.

### PR DESCRIPTION
**Pull request by Gastron**
Implements the functionality. Currently not tested at all.

### CODE REVIEW

File: speechbrain/utils/distributed.py [hunk 1]
```diff
@@ -51,25 +55,134 @@ def run_on_main(
...
+    Arguments
+    ---------
+    run_opts: list
+        a list of arguments to parse, most often from `sys.argv[1:]`
+    """
+    if run_opts["distributed_launch"]:
+        if "local_rank" not in run_opts:
+            raise ValueError(
+                "To use DDP backend, start your script with:\n\t"
+                "python -m torch.distributed.lunch [args]\n\t"
```
Comment (csukuangfj) [inline speechbrain/utils/distributed.py:125]: It is `launch`, not `lunch`. The same typo exists for the following comments.

File: speechbrain/utils/distributed.py [hunk 2]
```diff
@@ -51,25 +55,134 @@ def run_on_main(
...
+        else:
+            if run_opts["local_rank"] + 1 > torch.cuda.device_count():
+                raise ValueError(
+                    "Killing process " + str() + "\n"
+                    "To use DDP backend, start your script with:\n\t"
+                    "python -m torch.distributed.lunch [args]\n\t"
+                    "experiment.py hyperparams.yaml --distributed_launch=True "
+                    "--distributed_backend=nccl"
+                )
+        if "RANK" in os.environ is None or os.environ["RANK"] == "":
```
Comment (csukuangfj) [inline speechbrain/utils/distributed.py:138]: The return value of `"RANK" in os.environ`  is a boolean, which can never be `None`.

So
```python
"RANK" in os.environ is None
```
does not make sense.

File: speechbrain/utils/distributed.py (hunk 1 above)
Comment (TParcollet) [reply speechbrain/utils/distributed.py:125]: Solved, thanks.

File: speechbrain/utils/distributed.py (hunk 2 above)
Comment (TParcollet) [reply speechbrain/utils/distributed.py:138]: Where did you find that it is a boolean ? From what I see from the doc, it is a string.

Comment (csukuangfj) [reply speechbrain/utils/distributed.py:138]: It is from https://docs.python.org/3/reference/datamodel.html#object.__contains__

---

You can write a test
```python
>>> isinstance('' in {}, bool)
True
```

`os.environ` is a dict. 

`"RANK" in os.environ` will invoke the `__contains__` method of the dict, which returns `True` or `False`.
It never returns a string, I believe.

---

> From what I see from the doc, it is a string.

Could you point me to the doc you are reading?

Comment (TParcollet) [reply speechbrain/utils/distributed.py:138]: https://docs.python.org/3/library/os.html

@aheba

Comment (csukuangfj) [reply speechbrain/utils/distributed.py:138]: https://docs.python.org/3/library/os.html#os.environ
reads
> A mapping object representing the string environment.

which shows `os.environ` is indeed a dict, i.e., a mapping object. Cannot find where it says `os.environ` is a string or the result of
`"RANK" in os.environ`  is a string.

Comment (TParcollet) [reply speechbrain/utils/distributed.py:138]: Right, could you do a PR to fix this ? So we can test it :D

**Pull request merged**